constraint/pkg/client/e2e_test.go

```diff
@@ -0,0 +1,694 @@
+package client
+
+import (
+	"context"
+	"encoding/json"
+	"strings"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	"github.com/open-policy-agent/frameworks/constraint/pkg/client/drivers/local"
+	"github.com/open-policy-agent/frameworks/constraint/pkg/core/templates"
+	"github.com/open-policy-agent/frameworks/constraint/pkg/types"
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	k8schema "k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+const (
+	denied    = "DENIED"
+	rejection = "REJECTION"
+)
+
+func newConstraintTemplate(name, rego string, libs ...string) *templates.ConstraintTemplate {
+	return &templates.ConstraintTemplate{
+		ObjectMeta: metav1.ObjectMeta{Name: strings.ToLower(name)},
+		Spec: templates.ConstraintTemplateSpec{
+			CRD: templates.CRD{
+				Spec: templates.CRDSpec{
+					Names: templates.Names{
+						Kind: name,
+					},
+					Validation: &templates.Validation{
+						OpenAPIV3Schema: &apiextensions.JSONSchemaProps{
+							Type: "object",
+							Properties: map[string]apiextensions.JSONSchemaProps{
+								"expected": {Type: "string"},
+							},
+						},
+					},
+				},
+			},
+			Targets: []templates.Target{
+				{Target: "test.target", Rego: rego, Libs: libs},
+			},
+		},
+	}
+}
+
+func newConstraint(kind, name string, params map[string]string, enforcementAction *string) *unstructured.Unstructured {
+	c := &unstructured.Unstructured{}
+	c.SetGroupVersionKind(k8schema.GroupVersionKind{
+		Group:   "constraints.gatekeeper.sh",
+		Version: "v1alpha1",
+		Kind:    kind,
+	})
+	c.SetName(name)
+	if enforcementAction != nil {
+		if err := unstructured.SetNestedField(c.Object, *enforcementAction, "spec", "enforcementAction"); err != nil {
+			panic(err)
+		}
+	}
+	if err := unstructured.SetNestedStringMap(c.Object, params, "spec", "parameters"); err != nil {
+		panic(err)
+	}
+	return c
+}
+
+const (
+	// basic deny template.
+	denyTemplateRego = `package foo
+violation[{"msg": "DENIED", "details": {}}] {
+	"always" == "always"
+}`
+
+	// basic deny template that uses a lib rule.
+	denyTemplateWithLibRego = `package foo
+
+import data.lib.bar
+
+violation[{"msg": "DENIED", "details": {}}] {
+  bar.always[x]
+	x == "always"
+}`
+
+	denyTemplateWithLibLib = `package lib.bar
+always[y] {
+  y = "always"
+}
+`
+)
+
+var denyAllCases = []struct {
```

Comment:
Why not have positive cases as well?  I can imagine a scenario where a bug is introduced (perhaps in the rego library) that causes all templates be rejected.  These tests might still all/mostly pass.

Reply:
I agree. This PR is only for refactoring test cases so that adding such tests (which are indeed absolutely necessary) is easier. Having those in this PR would have made it too large.